databricks_cli/stack/api.py

```diff
@@ -0,0 +1,246 @@
+        if not os.path.exists(stack_file_folder):
+            os.makedirs(stack_file_folder)
+        with open(stack_status_filepath, 'w+') as f:
+            click.echo('Storing deploy status metadata to %s' % stack_status_filepath)
+            json.dump(data, f, indent=2)
+
+        if custom_path:
+            custom_path_folder = os.path.dirname(custom_path)
+            if not os.path.exists(custom_path_folder):
+                os.makedirs(custom_path_folder)
+            with open(custom_path, 'w+') as f:
+                click.echo('Storing deploy status metadata to %s' % os.path.abspath(custom_path))
+                json.dump(data, f, indent=2)
+
+    def deploy_job(self, resource_id, job_settings, physical_id=None):
+        job_id = None
+        print("Deploying job %s with settings: \n%s \n" % (resource_id, json.dumps(
+            job_settings, indent=2, separators=(',', ': '))))
+
+        if physical_id:  # job exists
+            job_id = physical_id
+        elif 'name' in job_settings:
+            jobs_same_name = self.jobs_client.get_jobs_by_name(job_settings['name'])
+            if jobs_same_name:
+                creator_name = jobs_same_name[0]['creator_user_name']
+                timestamp = jobs_same_name[0]['created_time'] / MS_SEC
+                date_created = datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
+                click.echo('Warning: Job exists with same name created by %s on %s. Job will '
+                           'be overwritten' % (creator_name, date_created))
+                job_id = jobs_same_name[0]['job_id']
+
+        if job_id:
+            try:
+                # Check if persisted job still exists, otherwise create new job.
+                self.jobs_client.get_job(job_id)
+            except HTTPError:
+                job_id = None
+
+        if job_id:
+            click.echo("Updating Job: %s" % resource_id)
+            self.jobs_client.reset_job({'job_id': job_id, 'new_settings': job_settings})
+            click.echo("Link: %s#job/%s" % (self.host, str(job_id)))
+        else:
+            click.echo("Creating Job: %s" % resource_id)
+            job_id = self.jobs_client.create_job(job_settings)['job_id']
+            click.echo("%s Created with ID %s. Link: %s#job/%s" % (
+                resource_id, str(job_id), self.host, str(job_id)))
+
+        deploy_output = self.jobs_client.get_job(job_id)
+
+        return job_id, deploy_output
+
+    def deploy_resource(self, resource):  # overwrite to be added
+        try:
+            resource_id = resource[RESOURCE_ID]
+            resource_type = resource[RESOURCE_TYPE]
+            resource_properties = resource[RESOURCE_PROPERTIES]
+        except KeyError as e:
+            raise ConfigError("%s doesn't exist in resource config" % str(e))
+
+        # Deployment
+        physical_id = self._get_deployed_resource(resource_id, resource_type)
+
+        if resource_type == JOBS_TYPE:
```

Comment:
`deploy_resource` should only do triage. For example, it could do
```
if resource_type == JOBS_TYPE:
  # This function figures out how to make use of resource_properties and physical_id
  deploy_job(resource_properties, physical_id)
```

Reply:
Basically `deploy_resource` should not know anything specific to any resource types.

Reply:
Makes sense. I was just worried that ‘deploy_job’ wouldnt have a clean interface so i handled Parsing fields of the inputted json here. I’m going to look into create functions that take in resource properties and physical id’s for a specific resource types.

Reply:
The advantage of my proposal is that you can put all the logic related to jobs (or dbfs etc) in one place (inside deploy_job). Otherwise, whenever you add a new component (or change the interface  for an exsiting one) you will need to make non-trivial changes in multiple places

Reply:
sounds good.